pkg/instancetype/compatibility/compatibility.go

```diff
@@ -0,0 +1,196 @@
+//nolint:dupl,lll,gocyclo
+package compatibility
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/json"
+
+	"kubevirt.io/api/instancetype/v1alpha1"
+	"kubevirt.io/api/instancetype/v1alpha2"
+	"kubevirt.io/api/instancetype/v1beta1"
+	generatedscheme "kubevirt.io/client-go/generated/kubevirt/clientset/versioned/scheme"
+)
+
+func GetInstancetypeSpec(revision *appsv1.ControllerRevision) (*v1beta1.VirtualMachineInstancetypeSpec, error) {
+	if err := Decode(revision); err != nil {
+		return nil, err
+	}
+	switch obj := revision.Data.Object.(type) {
+	case *v1beta1.VirtualMachineInstancetype:
+		return &obj.Spec, nil
+	case *v1beta1.VirtualMachineClusterInstancetype:
+		return &obj.Spec, nil
+	default:
+		return nil, fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
+	}
+}
+
+func GetPreferenceSpec(revision *appsv1.ControllerRevision) (*v1beta1.VirtualMachinePreferenceSpec, error) {
+	if err := Decode(revision); err != nil {
+		return nil, err
+	}
+	switch obj := revision.Data.Object.(type) {
+	case *v1beta1.VirtualMachinePreference:
+		return &obj.Spec, nil
+	case *v1beta1.VirtualMachineClusterPreference:
+		return &obj.Spec, nil
+	default:
+		return nil, fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
+	}
+}
+
+func Decode(revision *appsv1.ControllerRevision) error {
+	if len(revision.Data.Raw) == 0 {
+		return nil
+	}
+
+	// Backward compatibility check. Try to decode ControllerRevision from v1alpha1 version.
+	oldObject, err := decodeSpecRevision(revision.Data.Raw)
+	if err != nil {
+		return fmt.Errorf("failed to decode old ControllerRevision: %w", err)
+	}
+	if oldObject != nil {
+		revision.Data.Object = oldObject
+		return nil
+	}
+	return decodeControllerRevisionObject(revision)
+}
+
+func decodeControllerRevisionObject(revision *appsv1.ControllerRevision) error {
```

Comment:
there is a lot of code that repeats itself in this function... 
Do you think we can abstract this somehow?

Reply:
You just end up repeating the existing `switch` conditionals within `convert` with that approach.

I've created https://github.com/kubevirt/kubevirt/issues/12982 to track reworking this in the future as the aim of this refactor is more to make implementation of some final features such as flattening, runtime state stored in status and custom sizing easier to land in this release cycle. We can follow up with further refactors once that's done.

Reply:
Not exactly. This way you centralize the `convert` functionality and `decodeControllerRevisionObject` becomes easier to follow since it only handles type detection.